src/main/java/org/pac4j/vertx/handler/impl/CallbackHandler.java

```diff
@@ -34,75 +40,67 @@
-
-        redirectToTarget(routingContext, sessionAttributes);
-    }
-
-    private void redirectToTarget(final RoutingContext routingContext, final Pac4jSessionAttributes sessionAttributes) {
-
-      String requestedUrl = retrieveOriginalUrl(sessionAttributes);
-      saveUrl(null, sessionAttributes);
-      final String redirectUrl = defaultUrl(requestedUrl, BaseConfig.getDefaultSuccessUrl());
-      pac4jAuthProvider.saveSessionAttributes(routingContext, sessionAttributes);
-      HttpResponseHelper.redirect(routingContext.request(), redirectUrl);
-
+  public void handle(RoutingContext event) {
+
+    // Can we complete the authentication process here?
+    final VertxWebContext webContext = new VertxWebContext(event);
+    final ProfileManager profileManager = new ProfileManager(webContext);
+    final Client client = config.getClients().findClient(webContext);
+
+    CommonHelper.assertNotNull("client", client);
+
+    final Credentials credentials;
+    try {
+      credentials = client.getCredentials(webContext);
+    } catch (final RequiresHttpAction e) {
+      LOG.warn("Requires http action: " + e.getCode());
+      httpActionHandler.handle(e.getCode(), webContext);
+      return;
     }
-
-  /**
-   * This method returns the default url from a specified url compared with a default url.
-   *
-   * @param url a specific url
-   * @param defaultUrl the default url
-   * @return the default url
-   */
-  public static String defaultUrl(final String url, final String defaultUrl) {
-    String redirectUrl = defaultUrl;
-    if (url != null && !"".equals(url)) {
-      redirectUrl = url;
-    }
-    logger.debug("defaultUrl : {}", redirectUrl);
-    return redirectUrl;
+    vertx.<CommonProfile>executeBlocking(future -> {
+      final CommonProfile profile = (CommonProfile) client.getUserProfile(credentials, webContext);
+      future.complete(profile);
+    }, result -> {
+      if (result.succeeded()) {
+        Optional.ofNullable(result.result()).ifPresent(commonProfile -> {
+          profileManager.save(true, commonProfile);
+        });
+        redirectToOriginallyRequestedUrl(webContext);
+      } else {
+        event.fail(new RuntimeException("Failed to retrieve user profile"));
+      }
+    });
   }
 
-  protected String retrieveOriginalUrl(final Pac4jSessionAttributes sessionAttributes) {
-    return (String) sessionAttributes.getCustomAttributes().get(Pac4jConstants.REQUESTED_URL);
+  private void redirectToOriginallyRequestedUrl(final VertxWebContext webContext) {
+    final String requestedUrl = (String) webContext.getSessionAttribute(Pac4jConstants.REQUESTED_URL);
+    LOG.debug("requestedUrl: " + requestedUrl);
+    if (CommonHelper.isNotBlank(requestedUrl)) {
+      webContext.setSessionAttribute(Pac4jConstants.REQUESTED_URL, null);
+      webContext.setResponseStatus(302);
+      webContext.setResponseHeader("location", requestedUrl);
+      webContext.completeResponse();
+    } else {
```

Comment:
You mean that the setResponseStatus and completeResponse should be removed?

Keep in mind that we need the concept of completing a response for vert.x - you have to call response.end() on the underlying HttpServerResponse or it will not complete on the client. Where else would you suggest this should happen - it needs to be called somewhere.

> On 18 Sep 2015, at 10:49, LELEU Jérôme <notifications@github.com> wrote:
> 
> In src/main/java/org/pac4j/vertx/handler/impl/CallbackHandler.java <https://github.com/pac4j/vertx-pac4j/pull/21#discussion_r39839599>:
> 
> >    }
> >  
> > -  protected String retrieveOriginalUrl(final Pac4jSessionAttributes sessionAttributes) {
> > -    return (String) sessionAttributes.getCustomAttributes().get(Pac4jConstants.REQUESTED_URL);
> > +  private void redirectToOriginallyRequestedUrl(final VertxWebContext webContext) {
> > +    final String requestedUrl = (String) webContext.getSessionAttribute(Pac4jConstants.REQUESTED_URL);
> > +    LOG.debug("requestedUrl: " + requestedUrl);
> > +    if (CommonHelper.isNotBlank(requestedUrl)) {
> > +      webContext.setSessionAttribute(Pac4jConstants.REQUESTED_URL, null);
> > +      webContext.setResponseStatus(302);
> > +      webContext.setResponseHeader("location", requestedUrl);
> > +      webContext.completeResponse();
> > +    } else {
> I'm surprised by this code: no setResponseStatus, no completeResponse...
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/pac4j/vertx-pac4j/pull/21/files#r39839599>.
>